### PR TITLE
Autoloader

### PR DESCRIPTION
**Pull request by VioletRed**
Changes to reduce initialization time.

First change is not loading all resolvers into memory at init time, but only when they are used. To do this, metaclass AutoloaderMeta catches class initialization, and create wrappers for inherited functions that point  to an internal object self._ref. For example, class UrlWrapper, that inherits from UrlResolver, PluginSettings, SiteAuth & AutoloadPlugin, will have a method called get_media_url() inherited from UrlResolver, that would actually call self._ref.get_media_url(). By default, such methods just call "not_implemented()", which triggers an exception catched by the wrapper and used to load the actual resolver. The only requirement is that resolvers MUST have a name, but otherwise they are unchanged and unaware of any change in plugnplay library. Note that this change by itself doesn't change total resolution time, as all resolvers are still loaded to memory when instantiating HostMediaFile().

The second patch is simpler, as it only checks if there has been a software update before writing a new settings.xml file. It also adds some pretty printing to the file, and sorts the resolvers by name to make them easier to find in the settings screen.

### CODE REVIEW

**Comment by tknorris** [comment]
This looks much more useable. I have some comments that I will add on the commits.

File: lib/urlresolver/__init__.py [hunk 1]
```diff
@@ -175,19 +188,34 @@ def _update_settings_xml():
...
         finally:
             f.close
     except IOError:
         common.addon.log_error('error writing ' + common.settings_file)
 
-
-#make sure settings.xml is up to date
-_update_settings_xml()
+#Update settings.xml if newer plugin version
+if common.addon.get_setting('addon_version') != common.addon.get_version():
```
Comment (tknorris) [inline lib/urlresolver/__init__.py]: I don't think this will work the way you want because xbmc returns blank from get_setting if the setting name does not exist in the addon's settings.xml. The net result will be this will always be true and therefore update_settings_xml will always get called. In my addons when I do this kind of thing I directly compare the current settings.xml with the one I built and if they are different, replace it.

File: lib/urlresolver/plugnplay/interfaces.py [hunk 2]
```diff
@@ -66,14 +70,19 @@ class UrlResolver(Interface):
     '''
     
     name = 'override_me'
-    '''(str) A human readable name for your plugin.'''
+    '''(str) A human readable name for your plugin. Must match plugin filename'''
```
Comment (tknorris) [inline lib/urlresolver/plugnplay/interfaces.py]: Is this really true? Is there some way to avoid the requirement that the name match the filename?

File: lib/urlresolver/plugnplay/interfaces.py [hunk 3]
```diff
@@ -137,9 +147,9 @@ def get_url(self, host, media_id):
 
     def get_host_and_id(self, url):
         not_implemented(self)
-
-
-    def valid_url(self, web_url):
+    
+    
+    def valid_url(self, web_url, host):
```
Comment (tknorris) [inline lib/urlresolver/plugnplay/interfaces.py]: why the change in the valid_url function signature?

File: lib/urlresolver/__init__.py (hunk 1 above)
Comment (Eldorados) [inline lib/urlresolver/__init__.py]: @tknorris If I'm reading right, I will disagree

At line 221 addon_version is written to the settings, so on next initialization this will return false as intended since get_setting will no longer return a blank

If addon_version does not exist ie it has just been rebuilt from an update or other, then this will correctly return True and write the version to the xml

File: lib/urlresolver/plugnplay/interfaces.py (hunk 3 above)
Comment (Eldorados) [inline lib/urlresolver/plugnplay/interfaces.py]: Curious as well, simply to mirror the override used in the resolvers?

File: lib/urlresolver/__init__.py (hunk 1 above)
Comment (tknorris) [inline lib/urlresolver/__init__.py]: setSetting doesn't add it ot the settings.xml, it just sets the setting if it exists. In my development, I've found that it doesn't work. You can try it out by just throwing a setSetting() immediately followed by a getSetting() for a setting that doesn't exist. When I've tried this in the past that getSetting() call always returned blank.

Comment (Eldorados) [inline lib/urlresolver/__init__.py]: Ah yes your right, overlooked that part

'addon_version' should be set in _update_settings_xml() and even better with a default value containing current version

Comment (tknorris) [inline lib/urlresolver/__init__.py]: Well, this is interseting. I just tested this again on Helix and doing the set_setting for a non-existant setting actually persisted it into the userdata settings.xml and get_setting returned it. I know for a fact that didn't working Gotham and before because I needed it several times. I still think for pre-Helix compatibility we should have it in the settings.xml but it's interesting that it doesn't seem to be strictly required anymore.

File: lib/urlresolver/plugnplay/manager.py [hunk 4]
```diff
@@ -23,18 +23,45 @@
...
-                                                      implementor_instance)
-            return
-    self.iface_implementors[interface].append(implementor_instance)
 
-  def implementors(self, interface):
-    return self.iface_implementors.get(interface, [])
+
+    def add_implementor(self, interface, implementor_instance):
+        self.iface_implementors.setdefault(interface, [])
+        index = self.binary_search(self.iface_implementors[interface], implementor_instance)
```
Comment (tknorris) [inline lib/urlresolver/plugnplay/manager.py]: I'm not sure about the binary_search change. The old code would add the next resolver at the first index where the next resolvers priority was <= the one in the list. 90% of the time all the resolvers were set to 100 so that meant each new resolver was added in at index 0 (i.e. only one loop iteration). The only exception to this was when people set debrid resolvers to < 100 and then only those resolvers that were processed after the debrid ones would be put right after the debrid ones. No one runs more than 1 or 2 debrid resolvers. So, that means 90% of the time the old loop was one iteration, and rarely 2-3 iterations. Is the same thing true of the binray search? binary_search on average is faster than an in order search, but I think we have a special case here.

Comment (tknorris) [inline lib/urlresolver/plugnplay/manager.py]: It also just occured to me that binary search is always dependent on a sorted array. Sorting the list has be more costly than the old algorithm?

File: lib/urlresolver/plugnplay/interfaces.py (hunk 2 above)
Comment (VioletRed) [inline lib/urlresolver/plugnplay/interfaces.py]: No, that was a first try using "name" to match "filename", but it's not necessary anymore.

I'll fix the comment.

Comment (tknorris) [inline lib/urlresolver/plugnplay/interfaces.py]: Great!

File: lib/urlresolver/plugnplay/interfaces.py (hunk 3 above)
Comment (VioletRed) [inline lib/urlresolver/plugnplay/interfaces.py]: The signature didn't match the override, but the metaclass will use the signature to define the exposed API.

File: lib/urlresolver/plugnplay/manager.py (hunk 4 above)
Comment (VioletRed) [inline lib/urlresolver/plugnplay/manager.py]: Good question. I'd try keeping the old sorting and see how it affects, but you're probably right about not making any difference if the user doesn't sort the resolvers by hand.

File: lib/urlresolver/plugnplay/interfaces.py (hunk 3 above)
Comment (tknorris) [inline lib/urlresolver/plugnplay/interfaces.py]: That sounds like that is part of the domain resolution logic. Is that right? If so, it shouldn't be part of this commit.

File: lib/urlresolver/__init__.py (hunk 1 above)
Comment (VioletRed) [inline lib/urlresolver/__init__.py]: I think it works from Gotham, at least. I just copied the "first_run" trick they did in the old youtube plugin.

Comment (tknorris) [inline lib/urlresolver/__init__.py]: It definitely doesn't in Gotham. I spent half a day a long time ago trying to figure out why set_setting wasn't actually setting a setting and it was because it wasn't speciifed in settings.xml. I eventually found a post on the .org forum that it was required. You can add it the new "Urlresolver" category you created and just create it with visible="false" and it won't be seen in the front end.

Comment (VioletRed) [inline lib/urlresolver/__init__.py]: Ok, I will.

File: lib/urlresolver/plugnplay/interfaces.py (hunk 3 above)
Comment (VioletRed) [inline lib/urlresolver/plugnplay/interfaces.py]: It's part of this commit. The "attr_set_property" function, that sets the names of the methods, is only called when first defining a class that inherits "Autoloader", not when the actual plugin is loaded into memory. It means that when the plugin is loaded, python won't redefine the method signature, making it impossible to call overridden methods that have a different signature from the parent class.

Comment (tknorris) [inline lib/urlresolver/plugnplay/interfaces.py]: What I don't understand though is why is "host" added to the list of parameters? All the resolvers should only be taking the web_url into valid_url? I don't see how host would be used in valid_url until the domain checks are added in...

**Comment by VioletRed** [comment]
Just push updated head (addc6b63e827dec050dca6a2d78cac7fc949e18e) addressing tknorris comments.

**Comment by tknorris** [comment]
Looks pretty good to me. Do you have numbers now on how expensive importing urlresolver is now? Once this change is released, will it be fast enough to move the import the be global?

**Comment by VioletRed** [comment]
I don't have exact numbers because the timestamp has only second-resolution. But it is well under 1 second in the raspberry pi.

**Comment by Eldorados** [comment]
Things are looking good, I'll give one more day for comments until pulling in and will immediately send out a beta
